### PR TITLE
bugfix: Make the Azure Defender clause robust against a non-existent …

### LINKED ISSUE #257: Unable to set `log_analytics_workspace_enabled` to `false` and supply my own `log_analytics_workspace`
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Greenfield/Brownfield provisioning

greenfield

### Terraform Version

1.2.6

### Module Version

6.0.0

### AzureRM Provider Version

3.24.0

### Affected Resource(s)/Data Source(s)

azurerm_kubernetes_cluster

### Terraform Configuration Files

```hcl
# Resources 
locals {
  resource_group_name = "my-existing-resource-group"
}

data "azurerm_resource_group" "main_resource_group"{
    name = local.resource_group_name
}

resource "azurerm_log_analytics_solution" "analytics_solution" {

  solution_name         = "ContainerInsights"
  location              = data.azurerm_resource_group.main_resource_group.location
  resource_group_name   = local.resource_group_name
  workspace_resource_id = azurerm_log_analytics_workspace.analytics_workspace.id
  workspace_name        = azurerm_log_analytics_workspace.analytics_workspace.name

  plan {
    publisher = "Microsoft"
    product   = "OMSGallery/ContainerInsights"
  }
}

# The workspace we with to use with Microsoft Defender
resource "azurerm_log_analytics_workspace" "analytics_workspace" {
  name = "dev-workspace"
  resource_group_name =  data.azurerm_resource_group.main_resource_group.name
  location = data.azurerm_resource_group.main_resource_group.location
}


module "aks" {
  source                                  = "Azure/aks/azurerm"
  version                                 = "6.0.0"
  resource_group_name                     = local.resource_group_name
  # The Analytics Workspace we created above
  log_analytics_workspace                 = azurerm_log_analytics_workspace.analytics_workspace
  # Enable Microsoft Defender
  microsoft_defender_enabled              = true
  # Disable the automatic provisioning of an Analytics Workspace
  log_analytics_workspace_enabled         = false  analytics workspace



  #[...]
  # Other variables required by the module.
```

### Expected Behaviour

I would expect the cluster to get provisioned with Microsoft Defender enabled and piping data into my Analytics Workspace

### Actual Behaviour

Setting `log_analytics_workspace_enabled` to `false` disabled the creation of the `azurerm_log_analytics_workspace.main` defined inside the AKS module. 

This causes a [`coalesce` expression in the `microsoft_defender` clause](https://github.com/Azure/terraform-azurerm-aks/blob/master/main.tf#L156) to fail with an `invalid index` since it assumes the existence of this resource:

```
│ Error: Invalid index
│ 
│   on .terraform/modules/aks-cluster-setup.aks/main.tf line 148, in resource "azurerm_kubernetes_cluster" "main":
│  148:       log_analytics_workspace_id = coalesce(try(var.log_analytics_workspace.id, null), azurerm_log_analytics_workspace.main[0].id)
│     ├────────────────
│     │ azurerm_log_analytics_workspace.main is empty tuple
│ 
│ The given key does not identify an element in this collection value: the
│ collection has no elements.

```

### Steps to Reproduce

`terraform plan`

### Important Factoids

None

### References

Introduced by https://github.com/Azure/terraform-azurerm-aks/pull/232

### PR DESCRIPTION
**Pull request by gzur**
Fixes #257

## Changes proposed in the pull request:
Support setting `var.microsoft_defender_enabled` to `false` and supplying your own analytics workspace to Microsoft Defender.

Signed-off-by: Gissur Þórhallsson <gissur@skyvafnir.is>

### CODE REVIEW

**Comment by lonegunmanb** [comment]
Hi @gzur thanks a lot for opening this pr! Unfortunately we've got a bug in our CI pipeline code and I've just fixed it, would you please merge your branch with the latest master branch so I can re-run the failed job again? Thanks!

**Comment by lonegunmanb** [comment]
Btw I've reviewed your code, almost LGTM but I've got a new idea. How about adding a new `local` like this:

```hcl
locals {
  log_analytics_workspace = var.log_analytics_workspace_enabled ? (
    var.log_analytics_workspace == null ? {
      id   = azurerm_log_analytics_workspace.main[0].id
      name = azurerm_log_analytics_workspace.main[0].name
      } : {
      id   = var.log_analytics_workspace.id
      name = var.log_analytics_workspace.name
    }
    ) : {
    id   = null
    name = null
  }
}
```

Then we change all references to workspace id and name to this `local`, sounds good?

**Review by lonegunmanb — CHANGES_REQUESTED** [review]
Please merge with the latest master branch, thanks for your understanding.

**Comment by gzur** [comment]
> Btw I've reviewed your code, almost LGTM but I've got a new idea. How about adding a new `local` like this right behind the workspace resource:
> 
> ```terraform
> locals {
>   log_analytics_workspace = var.log_analytics_workspace_enabled ? (
>     var.log_analytics_workspace == null ? {
>       id   = azurerm_log_analytics_workspace.main[0].id
>       name = azurerm_log_analytics_workspace.main[0].name
>       } : {
>       id   = var.log_analytics_workspace.id
>       name = var.log_analytics_workspace.name
>     }
>     ) : {
>     id   = null
>     name = null
>   }
> }
> ```
> 
> Then we change all references to workspace id and name to this `local`, sounds good?

Sure thing. 

But on a related note: we are introducing some fairly complex behaviour here. Do we have a framework for testing this?

**Comment by lonegunmanb** [comment]
> > Good question! What about we modify `named_cluster` folder in `examples` to create this log analysis workspace in root module and pass it via `var.log_analytics_workspace` so our e2e tests can cover both cases?
> 
> Sounds fun. I always wanted to try out how infrastructure tests work.
> 
> Should I modify the named_cluster test/example or duplicate it?

You can modify `examples/named_cluster/main.tf` directly, our e2e test steps will apply every sub folders under `examples` to ensure that all our examples work.

**Comment by lonegunmanb** [comment]
> > Good question! What about we modify `named_cluster` folder in `examples` to create this log analysis workspace in root module and pass it via `var.log_analytics_workspace` so our e2e tests can cover both cases?
> 
> Sounds fun. I always wanted to try out how infrastructure tests work.
> 
> Should I modify the named_cluster test/example or duplicate it?

@gzur I had designed "unit tests" for our new CI pipeline but I haven't added it to this aks module yet. I'll add one to test this complex `local` after we've merged your pr, please stay tuned if you're interested in how to write quick and cheap tests for `locals`.

**Comment by gzur** [comment]
> I'll add one to test this complex local after we've merged your pr, please stay tuned if you're interested in how to write quick and cheap tests for locals.

Ok. I'll hold off on diving into the tests until after we've merged this.

File: main.tf
Comment (lonegunmanb) [inline main.tf:21]: Would an object with `null` field be better (the caller don't need to check whether this object is `null`)?

```hcl
) : {
  id       = null
  name = null
}

File: main.tf
Comment (lonegunmanb) [inline main.tf:178]: Since we have a `precondition` to prevent enable ms defender without enable log analytics workspace, could we simplify this line to the following snippet?:

```hcl
  for_each = var.microsoft_defender_enabled ? ["microsoft_defender"] : []
```

File: main.tf
Comment (lonegunmanb) [inline main.tf:220]: Would the following code be better?:

```hcl
condition     = !var.microsoft_defender_enabled || var.log_analytics_workspace_enabled
```

File: variables.tf
Comment (lonegunmanb) [inline variables.tf:276]: Could the description be more specific? Like: 

```hcl
description = "(Optional) Is Microsoft Defender on the cluster enabled? Requires `var.log_analytics_workspace_enabled` to be `true` to set this variable to `true`."
```

**Review by lonegunmanb — CHANGES_REQUESTED** [review]
Thanks @gzur almost LGTM, only a few issues.

File: main.tf
Comment (gzur) [reply main.tf:178]: Agreed.

File: variables.tf
Comment (gzur) [reply variables.tf:276]: I agree. This is the way.

File: main.tf
Comment (gzur) [inline main.tf:209]: This removes the sole usage of this variable, and now `tflint` is reporting it as `terraform_unused_declarations`.

So, I'm going to remove it.

File: main.tf
Comment (gzur) [reply main.tf:21]: But won't the caller just have to check whether `local.log_analytics_workspace.id/name` is `null` instead? 

I feel that this breaks the Principle of Least Astonishment (maybe my background in `python` is showing).

File: main.tf
Comment (lonegunmanb) [reply main.tf]: Or you can use the following expression:

```hcl
condition     = !(var.microsoft_defender_enabled && !var.log_analytics_workspace_enabled)
```

That'll be easier to understand, right?

File: main.tf
Comment (lonegunmanb) [reply main.tf:21]: In Terraform assign `null` to an argument is equal to omit this argument, in some cases the caller doesn't need to check whether the log analytics workspace id is `null` or not, if the caller just assigns it to an optional argument, in that case we can save a `null` check. I personally prefer this "`null` safe" style, but I think your point also make sense. Please allow me to have more discuss with other people, thanks for your understanding.

File: main.tf
Comment (lonegunmanb) [reply main.tf:209]: I would prefer to preserve `var.log_analytics_solution_id` in this expression because this variable was introduced at first place in case the caller want to use log analytics workspace and solution they brought here instead of creating one. Remove this variable will be considered as a breaking change and we cannot introduce a breaking change in a minor version upgrade, which means this pr can only be merged in v7.0.0.

In this case I would prefer the original expression, which is:

```hcl
count = var.log_analytics_workspace_enabled && var.log_analytics_solution_id == null ? 1 : 0
```

File: variables.tf
Comment (lonegunmanb) [inline variables.tf:231]: Please preserve this variable, or it'll be considered as a breaking change.

File: main.tf
Comment (lonegunmanb) [inline main.tf:23]: I think we can remove this comment, could we?

File: main.tf
Comment (lonegunmanb) [inline main.tf:177]: Please remove this blank line to preserve a consistent code style, thanks.

File: main.tf
Comment (lonegunmanb) [inline main.tf:220]: As you suggested, we can use the following expression to make it easier to understand:

```hcl
condition     = !(var.microsoft_defender_enabled && !var.log_analytics_workspace_enabled)
```

**Review by lonegunmanb — CHANGES_REQUESTED** [review]
Please allow me to have some discusses with others on the `null` object issue. Thanks for your understanding.

File: variables.tf
Comment (gzur) [reply variables.tf:231]: Sure thing. I'll add a `tflint-ignore: terraform_unused_declarations` before it so that the linter behaves.

File: main.tf
Comment (gzur) [reply main.tf:23]: That is just to help people parse what's going on, since the clause above is a bit involved. 

But sure, I can remove it.

File: variables.tf
Comment (lonegunmanb) [reply variables.tf:231]: If we preserver the reference to this variable in `azurerm_log_analytics_solution`'s `count` expression, we won't need this tflint annotation, right?

File: main.tf
Comment (lonegunmanb) [reply main.tf:23]: Appreciate for your comment, yes they do help people to understand. What I mean is, could we just remove the comment in line 23?

Comment (gzur) [reply main.tf]: Done

**Review by lonegunmanb — CHANGES_REQUESTED** [review]
Thanks @gzur for your updating, I think we'll be ready to test this pr once we've rollbacked the `count` expression for `azurerm_log_analytics_solution` to the original version, and remove the tflint annotation. Thanks for your patience!

**Comment by gzur** [comment]
> Thanks @gzur for your updating, I think we'll be ready to test this pr once we've rollbacked the `count` expression for `azurerm_log_analytics_solution` to the original version, and remove the tflint annotation. Thanks for your patience!

How about:
```
count = (local.create_analytics_workspace && var.log_analytics_solution_id == null) ? 1 : 0
```

so that we use our recently introduced control variable.

I feel that this preserves the [original intent of `var.log_analytics_solution_id`](https://github.com/Azure/terraform-azurerm-aks/pull/213#discussion_r916391971) while also making the module easier to reason about.

**Comment by lonegunmanb** [comment]
> > Thanks @gzur for your updating, I think we'll be ready to test this pr once we've rollbacked the `count` expression for `azurerm_log_analytics_solution` to the original version, and remove the tflint annotation. Thanks for your patience!
> 
> How about:
> 
> ```
> count = (local.create_analytics_workspace && var.log_analytics_solution_id == null) ? 1 : 0
> ```
> 
> so that we use our recently introduced control variable.
> 
> I feel that this preserves the [original intent of `var.log_analytics_solution_id`](https://github.com/Azure/terraform-azurerm-aks/pull/213#discussion_r916391971) while also making the module easier to reason about.

My apology, it was mid-night in my time zone because from Oct 1st to 7th it'll be national holiday in China and I'll be on vacation, so I'd like to close all remaining pull requests before that, apparently I've made a mistake. It should be:

```hcl
count = var.log_analytics_workspace_enabled && var.log_analytics_solution_id == null ? 1 : 0
```

I think you've just pointed out a blind point of mine, that is we need a new `local` named `local.create_analytics_solution` just like `local.create_analytics_workspace`. Then the `count` expression will be:

```hcl
count = local.create_analytics_solution ? 1 : 0
```

This `local.create_analytics_solution` will be:

```hcl
create_analytics_workspace = var.log_analytics_workspace_enabled && var.log_analytics_solution_id == null
```

How does that sound?

**Comment by lonegunmanb** [comment]
An update, I've opened a new issue #263 to improve this `var.log_analytics_solution_id` and I'll submit a new pr to improve that after we merge this one.

**Comment by gzur** [comment]
> My apology, it was mid-night in my time zone because from Oct 1st to 7th it'll be national holiday in China and I'll be on vacation, so I'd like to close all remaining pull requests before that, [...]

No worries. Let's see if we can't get it merged then :)

> [...] It should be:
> ```
> count = var.log_analytics_workspace_enabled && var.log_analytics_solution_id == null ? 1 : 0
> ```

I'm assuming you are referring to the `azurerm_log_analytics_solution.main` resource here - for which we are introducing the `create_analytics_workspace` control variable (per your suggestion above).

Right?

**Comment by gzur** [comment]
>[...] 
> we need a new local named `local.create_analytics_solution` just like `local.create_analytics_workspace.` 
[...]
> How does that sound?

That sounds great. I'll put it in.

**Comment by gzur** [comment]
@lonegunmanb - Done

File: main.tf
Comment (gzur) [reply main.tf:21]: did we reach a consensus on:
```
) : {
  id       = null
  name = null
}
```
vs. 
```
) : null
```
?

**Review by lonegunmanb — APPROVED** [review]
Thanks @gzur , nice job and I cannot ask for more. It is the last pr we're going to merge for v6.1.0. LGTM!

**Pull request merged**
